core/src/main/scala/kafka/log/Log.scala

```diff
@@ -181,25 +191,38 @@ class Log(val dir: File,
         // if its a log file, load the corresponding log segment
         val start = filename.substring(0, filename.length - LogFileSuffix.length).toLong
         val indexFile = Log.indexFilename(dir, start)
+        val timeIndexFile = Log.timeIndexFilename(dir, start)
         val segment = new LogSegment(dir = dir,
                                      startOffset = start,
                                      indexIntervalBytes = config.indexInterval,
                                      maxIndexSize = config.maxIndexSize,
                                      rollJitterMs = config.randomSegmentJitter,
                                      time = time,
                                      fileAlreadyExists = true)
+        if (indexFile.exists() && timeIndexFile.exists()) {
 
-        if(indexFile.exists()) {
           try {
-              segment.index.sanityCheck()
+            segment.index.sanityCheck()
           } catch {
             case e: java.lang.IllegalArgumentException =>
-              warn("Found a corrupted index file, %s, deleting and rebuilding index...".format(indexFile.getAbsolutePath))
+              warn(s"Found a corrupted index file, deleting ${timeIndexFile.getAbsolutePath}, " +
+                s"${indexFile.getAbsolutePath} and rebuilding index...")
               indexFile.delete()
+              timeIndexFile.delete()
               segment.recover(config.maxMessageSize)
           }
-        }
-        else {
+
+          try {
+            segment.timeIndex.sanityCheck()
+          } catch {
+            case e: java.lang.IllegalArgumentException =>
+              warn(s"Found a corrupted time index file, deleting ${timeIndexFile.getAbsolutePath}, " +
+                  s"${indexFile.getAbsolutePath} and rebuilding index...")
+              indexFile.delete()
+              timeIndexFile.delete()
+              segment.recover(config.maxMessageSize)
+          }
+        } else {
```

Comment:
Hmm, this means that when upgrading, we are forced to scan all existing log segments since they won't have the timeindex. Could we just create an empty timeindex file in this case? See my other comment in LogSegment on whether it's useful to always have a non-empty timeindex.